src/System.Buffers/src/System/Buffers/DefaultArrayPool.cs

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace System.Buffers
+{
+    internal sealed class DefaultArrayPool<T> : ArrayPool<T>
+    {
+        private const int MinimiumArraySize = 16;
+        private int _maxArrayLength;
+        private int _numberOfArraysPerBucket;
+        private bool _isTObjectType;    // cache this to only do the reflection call once
+        private ArrayPoolBucket<T>[] _buckets;
+        private int _rentedBufferCount;
+        private int _totalBufferCount;
+
+        public override int RentedBufferCount
+        {
+            get
+            {
+                return Volatile.Read(ref _rentedBufferCount);
+            }
+        }
+
+        public override int TotalBufferCount
+        {
+            get
+            {
+                return Volatile.Read(ref _totalBufferCount);
+            }
+        }
+
+        internal DefaultArrayPool(int maxLength, int arraysPerBucket)
+        {
+            // Our bucketing algorithm has a minimum length of 16
+            if (maxLength < MinimiumArraySize)
+                throw new ArgumentOutOfRangeException(nameof(maxLength), SR.Maximum_Array_Size_Under_Minimum_Supported_Value);
+            if (maxLength <= 0)
+                throw new ArgumentOutOfRangeException(nameof(maxLength));
+            if (arraysPerBucket <= 0)
+                throw new ArgumentOutOfRangeException(nameof(arraysPerBucket));
+
+            int maxBuckets = Utilities.SelectBucketIndex(maxLength);
+            _isTObjectType = typeof(T).GetTypeInfo().IsClass;
+            _maxArrayLength = maxLength;
+            _numberOfArraysPerBucket = arraysPerBucket;
+            _buckets = new ArrayPoolBucket<T>[maxBuckets + 1];
+            _rentedBufferCount = 0;
+            _totalBufferCount = 0;
+        }
+
+        public override T[] Rent(int length)
+        {
+            if (length <= 0)
+                throw new ArgumentOutOfRangeException(nameof(length));
+
+            Interlocked.Increment(ref _rentedBufferCount);
+
+            int index = Utilities.SelectBucketIndex(length);
+            if (index >= _buckets.Length)
+            {
+                Interlocked.Increment(ref _totalBufferCount);
+                return new T[length];
+            }
+            else
+            {
+                T[] buffer = null;
+
+                // Search for an array starting at the 'index' bucket. If the bucket
+                // is empty, bump up to the next higher bucket and try that one
```

Comment:
Opened issue #4911 to track the tracing / diagnostics work and assigned to myself